docs/debugging_guide.md

```diff
@@ -0,0 +1,362 @@
+
+<img src="images/debug_crash_asm.png" width="600px" />
+
+There are also some implicitly defined functions that don't literally appear in your source code (e.g. a built-in copy constructor or assignment operator). For a crash in any of these functions, the debugger may just show you the first line of the class:
+
+<img src="images/debug_crash_implicit.png" width="600px" />
+
+**Tracing Back to _Your_ Code**  
+To diagnose crashes where the debugger initially shows you library code, you need to find the nearest part of _your_ code that ultimately called the library functions.
+
+Let's consider again the case of a crash in an implicitly-defined function. The debugger first just shows us the top line of the class definition, since there's no code to show for implicitly-defined functions.
+
+<img src="images/debug_crash_implicit_full.png" width="800px" />
+
+In this case, check the call stack. First, observe that the segfault did occur in an implicitly-defined function, the built-in `Card::operator=` assignment operator for copying a `Card`.
+
+<img src="images/debug_crash_4.png" width="400px" />
+
+Now, click the next stack frame below to see the where that operator was used. It was in our `Pack` constructor. Taking a look at the _calling_ stack frame(s) allows me to see where the problem originated and is generally sufficient to figure out what part of _my code_ was responsible for causing the implicitly-defined function to crash. In the picture below, we can see that an (obviously) incorrect loop is attempting to copy `Card`s to out-of-bounds array elements.
+
+<img src="images/debug_crash_5.png" width="800px" />
+
+## Breakpoints
+
+### Basic Usage
+
+<div class="primer-spec-callout info icon-info" markdown="1">
+A **breakpoint** pauses the program whenever it reaches a certain line.
+</div>
+
+For example, if you're running the main driver for a statistics project and your dataset summary doesn't print out correctly, you might want to pause the program after the summary has been calculated but before printing it out.
+
+To set a breakpoint, click to the left of a line as shown. A red dot appears to indicate the breakpoint on that line. (You can click again to unset it.)
+
+<img src="images/debug_breakpoint_0.png" width="700px" />
+
+Now, when I run the program via my debugger, it pauses just **before** running that line. That is, the highlighted line is up next, but has not run yet.
+
+<img src="images/debug_breakpoint_1.png" width="700px" />
+
+Now, I can inspect the `summary` variable and see if my `summarize()` function was returning the right result.
+
+<div class="primer-spec-callout warning" markdown="1">
+If your debugger isn't respecting your breakpoints or the lines where it pauses don't seem to match your source code, double check that you've **saved** and **re-compiled** any source files you were editing. (Some IDEs may auto-save or auto-build when you launch the debugger, but not all.)
```

Comment:
Another pitfall here is `-g` and not `-O...`.  Optimization makes weird things happen.  I could image 281 students referring back to this tutorial, which is why I mention it.

Reply:
Good call. Added this and others in db40345.